File: mobile/src/shared/lib/loop/supabaseLoopStore.ts
import { supabase } from '@/shared/api/client';
import type {
  AppendTurnInput,
  LoopRun,
  LoopStore,
  LoopTurn,
  OpenRunInput,
  RunStatus,
} from '@/shared/lib/loop/types';

function mapRun(row: Record<string, unknown>): LoopRun {
  return {
    id: String(row.id),
    goal: String(row.goal),
    status: row.status as LoopRun['status'],
    trigger: row.trigger as LoopRun['trigger'],
    ownerUserId: String(row.owner_user_id),
    goalDate: String(row.goal_date),
    maxIterations: Number(row.max_iterations ?? 10),
    maxWallClockMs: Number(row.max_wall_clock_ms ?? 0),
    createdAt: String(row.created_at),
    endedAt: row.ended_at ? String(row.ended_at) : null,
    endedReason: row.ended_reason ? String(row.ended_reason) : null,
  };
}

/** Supabase 구현의 LoopStore 포트 */
export function createSupabaseLoopStore(): LoopStore {
  return {
    async openRun(input: OpenRunInput): Promise<LoopRun> {
      const goal = `verify_day:${input.ownerUserId}:${input.goalDate}`;
      const { data, error } = await supabase
        .from('runs')
        .insert({
          goal,
          status: 'running',
          trigger: input.trigger,
          owner_user_id: input.ownerUserId,
          goal_date: input.goalDate,
          max_iterations: input.bounds.maxIterations,
          max_wall_clock_ms: input.bounds.maxWallClockMs,
        })
        .select('*')
        .single();

      if (error || !data) {
        throw new Error(error?.message ?? 'openRun failed');
      }
      return mapRun(data);
    },

    async getOpenRunForDay(ownerUserId, goalDate) {
      const { data, error } = await supabase
        .from('runs')
        .select('*')
        .eq('owner_user_id', ownerUserId)
        .eq('goal_date', goalDate)
        .eq('status', 'running')
        .order('created_at', { ascending: false })
        .limit(1)
        .maybeSingle();

      if (error) throw new Error(error.message);
      return data ? mapRun(data) : null;
    },

    async appendTurn(input: AppendTurnInput): Promise<LoopTurn> {
      const { data, error } = await supabase
        .from('turns')
        .insert({
          run_id: input.runId,
          turn: input.turn,
          plan_json: input.plan,
          result_json: input.result,
          observe_json: input.observe,
          verify_status: input.verifyStatus,
        })
        .select('*')
        .single();

      if (error || !data) {
        throw new Error(error?.message ?? 'appendTurn failed');
      }

      return {
        id: String(data.id),
        runId: String(data.run_id),
        turn: Number(data.turn),
        planJson: data.plan_json,
        resultJson: data.result_json,
        observeJson: data.observe_json,
        verifyStatus: data.verify_status,
        createdAt: String(data.created_at),
      };
    },

    async finishRun(runId, status: Exclude<RunStatus, 'running'>, endedReason) {
      const { error } = await supabase
        .from('runs')
        .update({
          status,
          ended_reason: endedReason,
          ended_at: new Date().toISOString(),
        })
        .eq('id', runId);

      if (error) throw new Error(error.message);
    },

    async listRecentObserveHashes(runId, limit) {
      const { data, error } = await supabase
        .from('turns')
        .select('observe_json')
        .eq('run_id', runId)
        .order('turn', { ascending: false })
        .limit(limit);

      if (error) throw new Error(error.message);
      return (data ?? []).map((row) => {
        const obs = row.observe_json as { pendingHash?: string } | null;
        return obs?.pendingHash ?? '';
      });
    },
  };
}

File: android-app/app/src/main/kotlin/com/jinlabs/yakok/data/SupabaseLoopStore.kt
package com.jinlabs.yakok.data

import com.jinlabs.yakok.core.med.DayLoop
import com.jinlabs.yakok.core.med.LoopRun
import com.jinlabs.yakok.core.med.LoopTrigger
import com.jinlabs.yakok.core.med.RunStatus
import com.jinlabs.yakok.core.med.VerifyStatus
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Order
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

@Serializable
data class RunRow(
    val id: String,
    val goal: String,
    val status: String,
    val trigger: String,
    @SerialName("owner_user_id") val ownerUserId: String? = null,
    @SerialName("goal_date") val goalDate: String? = null,
    @SerialName("max_iterations") val maxIterations: Int? = null,
    @SerialName("max_wall_clock_ms") val maxWallClockMs: Long? = null,
    @SerialName("created_at") val createdAt: String,
    @SerialName("ended_at") val endedAt: String? = null,
    @SerialName("ended_reason") val endedReason: String? = null,
)

@Serializable
data class ObserveRow(@SerialName("observe_json") val observeJson: JsonElement? = null)

@Singleton
class SupabaseLoopStore @Inject constructor(
    private val supabase: SupabaseClient,
) {
    suspend fun openRun(
        ownerUserId: String,
        goalDate: String,
        trigger: LoopTrigger,
        maxIterations: Int,
        maxWallClockMs: Long,
    ): LoopRun {
        val row = supabase.from("runs").insert(
            buildJsonObject {
                put("goal", kotlinx.serialization.json.JsonPrimitive("verify_day:$ownerUserId:$goalDate"))
                put("status", kotlinx.serialization.json.JsonPrimitive("running"))
                put("trigger", kotlinx.serialization.json.JsonPrimitive(DayLoop.triggerWire(trigger)))
                put("owner_user_id", kotlinx.serialization.json.JsonPrimitive(ownerUserId))
                put("goal_date", kotlinx.serialization.json.JsonPrimitive(goalDate))
                put("max_iterations", kotlinx.serialization.json.JsonPrimitive(maxIterations))
                put("max_wall_clock_ms", kotlinx.serialization.json.JsonPrimitive(maxWallClockMs))
            },
        ) { select() }.decodeSingle<RunRow>()
        return row.toLoopRun()
    }

    suspend fun getOpenRunForDay(ownerUserId: String, goalDate: String): LoopRun? =
        supabase.from("runs").select {
            filter {
                eq("owner_user_id", ownerUserId)
                eq("goal_date", goalDate)
                eq("status", "running")
            }
            order("created_at", Order.DESCENDING)
            limit(1)
        }.decodeList<RunRow>().firstOrNull()?.toLoopRun()

    suspend fun appendTurn(
        runId: String,
        turn: Int,
        plan: JsonObject,
        result: JsonObject,
        observe: JsonObject,
        verifyStatus: VerifyStatus,
    ) {
        supabase.from("turns").insert(
            buildJsonObject {
                put("run_id", kotlinx.serialization.json.JsonPrimitive(runId))
                put("turn", kotlinx.serialization.json.JsonPrimitive(turn))
                put("plan_json", plan)
                put("result_json", result)
                put("observe_json", observe)
                put("verify_status", kotlinx.serialization.json.JsonPrimitive(verifyStatus.name.lowercase().let {
                    when (verifyStatus) {
                        VerifyStatus.Success -> "success"
                        VerifyStatus.Continue -> "continue"
                        VerifyStatus.FailedVerify -> "failed_verify"
                    }
                }))
            },
        )
    }

    suspend fun finishRun(runId: String, status: RunStatus, endedReason: String) {
        supabase.from("runs").update(
            buildJsonObject {
                put("status", kotlinx.serialization.json.JsonPrimitive(DayLoop.statusWire(status)))
                put("ended_reason", kotlinx.serialization.json.JsonPrimitive(endedReason))
                put("ended_at", kotlinx.serialization.json.JsonPrimitive(kotlinx.datetime.Clock.System.now().toString()))
            },
        ) { filter { eq("id", runId) } }
    }

    suspend fun countTurns(runId: String): Int =
        supabase.from("turns").select {
            filter { eq("run_id", runId) }
        }.decodeList<ObserveRow>().size

    suspend fun listRecentObserveHashes(runId: String, limit: Int): List<String> =
        supabase.from("turns").select {
            filter { eq("run_id", runId) }
            order("turn", Order.DESCENDING)
            limit(limit.toLong())
        }.decodeList<ObserveRow>().map { row ->
            val obj = row.observeJson as? JsonObject ?: row.observeJson?.jsonObject
            obj?.get("pendingHash")?.jsonPrimitive?.contentOrNull.orEmpty()
        }
}

fun RunRow.toLoopRun(): LoopRun = LoopRun(
    id = id,
    goal = goal,
    status = DayLoop.statusFromWire(status),
    trigger = DayLoop.triggerFromWire(trigger),
    ownerUserId = ownerUserId.orEmpty(),
    goalDate = goalDate.orEmpty(),
    maxIterations = maxIterations ?: DayLoop.MaxIterations,
    maxWallClockMs = maxWallClockMs ?: 0L,
    createdAt = createdAt,
    endedAt = endedAt,
    endedReason = endedReason,
)
